pkg/policy/query.go

```diff
@@ -67,17 +78,35 @@ func (q *ResourcesQueryCache) impl(bctx topdown.BuiltinContext, operands []*ast.
 		q.queriedResourceTypes[query.ResourceType] = struct{}{}
 	}
 
-	resources, err := q.ResolveResources(bctx.Context, query)
+	// Construct cache key
+	cacheKeyBytes, err := json.Marshal(query)
 	if err != nil {
 		return nil, err
 	}
+	cacheKey := string(cacheKeyBytes)
 
+	// Try cache first
+	q.cacheMutex.RLock()
+	cached, ok := q.cacheTerms[cacheKey]
+	q.cacheMutex.RUnlock()
+	if ok {
+		return cached, nil
+	}
+
+	// Use resolver
+	q.cacheMutex.Lock()
+	defer q.cacheMutex.Unlock()
+	resources, err := q.ResolveResources(bctx.Context, query)
```

Comment:
Because we share that map through the whole app, couldn't it be that this mutex doesn't appropriately lock?
E.g. that other instances of `ResourcesQueryCache` use the same map internally, but obviously not the same mutex?

Reply:
I'm not sure which `map` you mean so I'll just answer for both:

1. `cacheTerms map[string]*ast.Term` is protected by the mutex.  We do use this map from multiple threads, but we also use a pointer to the same mutex (see `func (q *ResourcesQueryCache) trackResourceTypes`) where we copy that mutex pointer.
2. `queriedResourceTypes map[string]struct{}` is where we track the resource types that a policy uses.  Because we do this per policy, we have a separate map per thread and we don't need to lock.  This makes a bunch of assumptions about how the code works though, so maybe it would be better to lock there as well?